File: Dynamic Programming/Maximum Subarray.cpp
class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        int n = nums.size();
        vector<int> dp(n);//dp[i] means the maximum subarray ending with A[i];
        dp[0] = nums[0];
        int res = dp[0];
        
        for(int i = 1; i < n; i++){
            dp[i] = nums[i] + (dp[i - 1] > 0 ? dp[i - 1] : 0);
            res = max(res, dp[i]);
        }
        
        return res;
    }
};

File: Kadane's/Maximum Subarray.rs
impl Solution {
    pub fn max_sub_array(nums: Vec<i32>) -> i32 {
        let mut last_sum = 0;
        let mut res = nums[0];

        for num in nums {
            last_sum += num;
            res = res.max(last_sum);
            if last_sum < 1 {
                last_sum = 0;
            }
        }
        res
    }
}